File: Python/56_Merge_Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:

        # sort the arrays
        # [[1,3],[3,6],[4,9],[8,10],[15,18][16,27]]
       # [[1,3],[2,6],[8,10],[15,18]]
    
        sorted_intervals = sorted(intervals, key=lambda lst: lst[0])
        result:List[List[int]] = []
            
        for i in range(len(sorted_intervals)):
            if len(result) == 0:
                result.append(sorted_intervals[i]) # [1,3]
            elif self.is_overlap(result[-1], sorted_intervals[i]): 
                 result[-1] = self.merge_sorted_overlapping_intervals(result[-1], sorted_intervals[i])
            else:
                result.append(sorted_intervals[i])
        return result
    
    def is_overlap(self, first:List[int], second:List[int] )->bool:
        return first[1] >= second[0]
    
    def merge_sorted_overlapping_intervals(self, first:List[int], second:List[int] )->List[int]:
        return [first[0], max(first[1], second[1])]

File: Java/56_merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        // 1: sort array first,
        // 2: keep cur start and cur end
        // 3: take interval, if it's overwrap, ( cur start < interval start)
        // keep the curstart and compare ends
        // 4: if it's not overwrap, start new interval for output

        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        // uses LinkedList as easy to operate the end
        // also we don't know the final length. so can't use int[][] in this stage
        LinkedList<int[]> merged = new LinkedList<>();
        for(int[] interval : intervals){
            // overwrap?
            if(merged.isEmpty() || merged.getLast()[1] < interval[0]){
                merged.(interval);
            }else{
                merged.getLast()[1] = Math.max(merged.getLast()[1], interval[1]);
            }
        }
        return merged.toArray(new int[merged.size()][]);
        
    }
}